worker/cache.go

```diff
@@ -0,0 +1,231 @@
+package worker
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.sia.tech/renterd/api"
+	"go.sia.tech/renterd/webhooks"
+)
+
+const (
+	cacheKeyDownloadContracts = "downloadcontracts"
+	cacheKeyGougingParams     = "gougingparams"
+
+	cacheEntryExpiry = 5 * time.Minute
+)
+
+type memoryCache struct {
+	items map[string]cacheEntry
+	mu    sync.RWMutex
+}
+
+type cacheEntry struct {
+	value  interface{}
+	expiry time.Time
+}
+
+func newMemoryCache() *memoryCache {
+	return &memoryCache{
+		items: make(map[string]cacheEntry),
+	}
+}
+
+func (c *memoryCache) Get(key string) (interface{}, bool) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	entry, ok := c.items[key]
+	if !ok || time.Now().After(entry.expiry) {
+		return nil, false
+	}
+	return entry.value, ok
+}
+
+func (c *memoryCache) Set(key string, value interface{}) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.items[key] = cacheEntry{
+		value:  value,
+		expiry: time.Now().Add(cacheEntryExpiry),
+	}
+}
+
+func (c *memoryCache) Invalidate(key string) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	delete(c.items, key)
+}
+
+type cache struct {
+	b     Bus
+	cache *memoryCache
+}
+
+func (w *worker) initCache() {
+	if w.cache != nil {
+		panic("cache already initialized") // developer error
+	}
+	w.cache = newCache(w.bus)
+}
+
+func newCache(b Bus) *cache {
+	return &cache{
+		b:     b,
+		cache: newMemoryCache(),
+	}
+}
+
+func (c *cache) DownloadContracts(ctx context.Context) ([]api.ContractMetadata, error) {
+	value, ok := c.cache.Get(cacheKeyDownloadContracts)
+	if ok {
+		return value.([]api.ContractMetadata), nil
+	}
+	contracts, err := c.b.Contracts(ctx, api.ContractsOpts{})
+	if err != nil {
+		return nil, err
+	}
+	c.cache.Set(cacheKeyDownloadContracts, contracts)
+	return contracts, nil
+}
+
+func (c *cache) GougingParams(ctx context.Context) (api.GougingParams, error) {
+	value, ok := c.cache.Get(cacheKeyGougingParams)
+	if ok {
+		return value.(api.GougingParams), nil
+	}
+
+	gp, err := c.b.GougingParams(ctx)
+	if err != nil {
+		return api.GougingParams{}, err
+	}
+	c.cache.Set(cacheKeyGougingParams, gp)
+	return gp, nil
+}
+
+func (c *cache) handleEvent(event webhooks.Event) error {
```

Comment:
Methods like this that are called outside the `cache` can be uppercase. I'd even consider moving this to `internal/worker/cache.go`.